### PR TITLE
Layer-wise Predictive Coding Inference Refactor with Compute-Aware Tuning and Stable Optimization

### PR DESCRIPTION
**Pull request by yohannes4321**
PR Description 
 
This PR refactors the predictive coding (PC) inference flow in PCTransformer, introduces layer-specific inference depth tuning, and improves the robustness and stability of the Bayesian optimization and training pipeline.

The changes focus on architectural clarity, better hyperparameter control, and preventing invalid optimization states from corrupting best-trial selection.
image

tuning file 70 trails 
[tuning.pdf](https://github.com/user-attachments/files/25813663/tuning.pdf)

### Generate best simmilar Text to the Document with smaller  0.6 Free Energy and Less than 700 PPL
`

[WARNING] FlashAttention is not installed. Falling back to standard attention.
[INFO] NumExpr defaulting to 2 threads.
[Rank 0] Using device: cuda:0
Model loaded successfully from checkpoints/final_model.pt

[Sample 1]
[GENERATED]: should IS hel now ors pher happ QUE sub unto say OKE way DW MER fellow ace Have N ve first llow woman IET against friend HENRY yet D DUCH qu pi Eng Who men Here just WARWICK think il ent NE ct fellow MER of K COR ange oun rown ast death vant tra SICINIUS As ard But bes vo lt yet AN said FR lit CA r duke would que gone out INIUS Warwick young ith great ra eye INIUS OR ison land u hy ENCE Romeo ince gh Go Thy Sir ors am Un pher MER unto tend forth ers death hither An pp ENIUS Clarence -- cious now head ase WARWICK thought wer dear bring Have ure those NTES IS most ase Warwick or All hate UL Have heaven father ted wn die ct MER set state MER ay mi cha stay u sorrow duke ab Ri he ser grace UL QUEEN II though Cl up INIUS ear jo x AN sor ers DUCHESS against cour Shall ENIUS sla the IET ister off N He r cha mean WARWICK U foo INC gh MER FR sel body DW Rome ISABELLA IS CAPUL bes ra DUCH

[Sample 2]
[GENERATED]: IET LAN mad OF now res one our ken ci ble should pur pi w pher um f back ence would OR th Th ace gh ke think bes come way DUCH ite IUS far ity tong why vant que may b With No BUCKINGHAM P hel WARWICK ins On arm those vo news ark ser wer pher He ear ef tong coun pher i how pray hither Romeo ten er great gh KE IET GAR would ber ap ast Clarence lords UT vy le To can HENRY ABELLA Do son tong ES none OF doth morrow ELIZ tong Sir HENRY eng tle KING heard ENTIO jo ELIZABETH uch mor said H or EDW great friend iti jest oul ace vo mad eye Warwick WARWICK ser cle gh MER old mad MER Edward te ves IET kind ay All How LAN wick land back bet char since UC ford ison l hy good U morrow WARWICK hope WARWICK shi f friend r ORD one Sir own ent pr AN Who against hen OF $ now si red So still ans g son put L GAR set bring cousin P would whose C bet f deed gone KINGHAM comes

`
### **1. PCTransformer Forward Pass Refactor and Faster Convergence**

Overview
The global inference loop controlled by config.T has been removed across the codebase.

Removal of global config.T:
[8a38006](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/8a380068990fdc66a2fde581957632911a25fa06)

Instead, inference is now executed sequentially per layer, with each component using its own dedicated iteration limit defined in the configuration.

This refactor enables layer-wise predictive coding inference, providing greater flexibility and improved experimental control.

New Layer-Specific Inference Parameters

Each layer now has an independent inference depth parameter:

Commit:
[151d39d](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/151d39dc82480ccdb8061ae019ec3c5e443b4bc3)

- embed_T → self.embedding.pc_layer
- attn_T → block.attn.pc_qkv
- linear_attn_T → block.attn.pc_output
- fc1_T → block.mlp.pc_layer1
- fc2_T → block.mlp.pc_layer2
- linear_output_T → self.output.pc_layer

Impact

- Removes architectural coupling caused by the global config.T
- Enables fine-grained experimental control
- Allows tuning to discover optimal inference depth allocation per layer
- Improves clarity and transparency of the inference scheduling mechanism
- Contributes to faster convergence with fewer training epochs

### **2. Compute-Aware Optimization for Faster Inference**

Commit:
[efaf77b](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/efaf77b18e77be885676dd5e3f780b42318c1a55)

A compute-aware objective function was introduced:

`energy + lambda_compute * inference_cost`
Implemented via:

`compute_inference_cost(config)`

This objective encourages configurations with lower inference depth when model quality is similar, resulting in:

**Faster inference
Reduced computational cost
Maintained output quality**

3. Optuna Search Space Update

Commit:
[34a7a40](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/34a7a405d40606dc3a8dae8d2e3544c7e7b84ff8)

The single global T hyperparameter has been removed from the tuning search space.
It has been replaced with layer-specific inference parameters:

```
embed_T = trial.suggest_int("embed_T", 3, 12, log=True)
attn_T = trial.suggest_int("attn_T", 3, 12, log=True)
linear_attn_T = trial.suggest_int("linear_attn_T", 3, 12, log=True)
fc1_T = trial.suggest_int("fc1_T", 3, 12, log=True)
fc2_T = trial.suggest_int("fc2_T", 3, 12, log=True)
linear_output_T = trial.suggest_int("linear_output_T", 3, 12, log=True)
```

This allows the optimizer to determine where additional inference depth is beneficial, instead of enforcing a uniform number of iterations across all layers.

4. Safer Result Logging

Commit:
[a0aea1a](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/a0aea1ab27d8e2d14942735293f710fd177c0288)

- Logging mechanisms were improved to increase robustness during tuning.
- Improvements
- Added safe numeric formatting
- Handles missing or invalid trial attributes gracefully
- Prevents crashes during final result logging

5. Configuration Simplification

Commit:
[c2851a7](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/126/commits/c2851a79e2abe8b4a8245431a8383f2683ce062f)

The following global parameters were removed:

- config.T
- alpha

They have been replaced with explicit layer-wise parameters:

- embed_T
- attn_T
- linear_attn_T
- fc1_T
- fc2_T
- linear_output_T
- lambda_output for speed

This change simplifies configuration management and aligns the implementation with the layer-wise inference design.

### CODE REVIEW

**Comment by Saad-Musema** [comment]
@yohannes4321 I noticed that you defined this function

 `def combined_loss(energy, compute_cost=0.0, lambda_compute=0.0):
    return float(energy + (lambda_compute * compute_cost))`
    
Also added both **lambda compute** and **compute cost** to the tuning. What exactly is the meaning of this? If it is to try and push down the number of T steps taken, I believe it is better to give them a smaller hyperparameter search space rather than adding two new tunable parameters. And even if it is done this way, do you think it is fair to consider the steps T taken in attention and in linear layer have the same amount of compute cost?


You also removed our combined_loss function all together(the one we used for computing a combined loss of perplexity and energy for dual optimization). Could you explain why that was done?

**Comment by Saad-Musema** [comment]
In the GPT config file, if you notice the structure of the file, we are trying not to initialize any variable there, since we are using it as a class with no defined variables, could you update your code to make it that way please?

**Comment by Saad-Musema** [comment]
```
    print("\nModel configuration parameters for this training run:")
    for k in [
        "vocab_size", "block_size", "peak_learning_rate", "warmup_steps", "n_embed",
        "dropout", "lr", "embed_T", "attn_T", "linear_attn_T", "fc1_T", "fc2_T", "linear_output_T",
        "num_heads", "n_blocks", "batch_size", "num_epochs", "update_bias",
        "internal_energy_fn_name", "output_energy_fn_name", "combined_internal_weight",
        "combined_output_weight", "use_flash_attention"
    ]:
        print(f"{k}={best_config.get(k)}")
   
```


Update this to use the logger that was already defined, please don't add unnecessary print statements.

**Comment by yohannes4321** [comment]
ok i will review it

**Comment by yohannes4321** [comment]
if you can can we make google meeting on this and i am implementing
research paper in code that you told me and to update you the best model
faster correct model

**Comment by yohannes4321** [comment]
@ 
Good Afternoon 

i have removed the lambda_cost and compute_cost from tuning and training and combined_loss function and also in GPT Config file i donot initialize any variable and i also removed print statement and only use Logger that is already defined 
 and i tune it 70 trails and using the best trail i train the model and i generate text on this new code 
 
 
and as combined loss it return energy because i talk to nardose and she told me to do so because 
before we use alpha *energy + 1-alpha *ppl this will use both  ppl and energy to minimize that is good but in this layer wise embed_T,atten_T,atten_out_T,fc1_T,fc2_T,linear_output_T  so when combined loss used both ppl and energy it make ppl so decrease too small and as i send to ppl come to 3 which makes the energy to increase and generate text will be bad

so the correct apporch is to use combined loss to return energy if we tune the best energy the smallest energy it will automatically the Perpelixity will decrase because the model allways decrease ppl 

generate text is the best text

**Comment by yohannes4321** [comment]
can you check it the code base please i have changed as what you told me
